openssl: delete cgo-less experiment and promote to default (#2106)
microsoft/go: eng/doc/NocgoOpenSSL.md

> # Cgo-less OpenSSL Backend

Reviewer: Fairly different from the filename? `CgolessOpenSSL.md` or:

```suggestion
# No-cgo OpenSSL Backend
```

> This document describes the cgo-less OpenSSL backend for the Microsoft build of Go, available as an experimental feature starting in Go 1.26.

Reviewer: This only really describes how to turn it on, not how it works (which people will naturally be curious/cautious of, so we should do what we can to point people in the right direction IMO). Should we have docs in the backend repo maybe? Also for the macOS nocgo backend?

## Overview

> In Go 1.26, there is a cgo-less experiment available for Linux: `ms_nocgo_opensslcrypto`.
> 
> In Go 1.27, the experiment will be removed and the cgo requirement for `systemcrypto` on Linux will be lifted by default.

Reviewer: ```suggestion
In Go 1.26, there is a cgo-less experiment available for Linux: `ms_nocgo_opensslcrypto`.

> [!NOTE]
> While `systemcrypto` is a fully supported `GOEXPERIMENT` value (it is not "experimental"), `ms_nocgo_opensslcrypto` **is** experimental in Go 1.26 and may have limitations.

In Go 1.27, the experiment will be removed and the cgo requirement for `systemcrypto` on Linux will be lifted by default.
```

This allows the use of OpenSSL without requiring cgo.

> ## Supported Architectures

Reviewer: ```suggestion
## Supported architectures
```

We generally aren't using title case for subheaders in the repo.

Currently this experiment is supported on the following architectures:

> - **386**
> - **amd64**
> - **arm**
> - **arm64**
> - **ppc64le**
> - **riscv64**

Reviewer: Like https://go.dev/dl/, I think we should highlight the ones that we expect most people to use, and therefore the ones they're probably looking for. (Otherwise, why highlight anything?)

```suggestion
- 386
- **amd64**
- arm
- **arm64**
- ppc64le
- riscv64
```

The set of supported architectures is limited because each architecture requires a unique assembly implementation to call OpenSSL.
Architectures are added based on demand and available resources.

> To see existing requests or request support for additional architectures, search for issues with the [`Area-Nocgo`](https://github.com/microsoft/go/labels/Area-Nocgo) label.

Reviewer: Prettier:

```suggestion
To see existing requests or request support for additional architectures, use the [![](https://img.shields.io/github/labels/microsoft/go/Area-Nocgo)](https://github.com/microsoft/go/labels/Area-Nocgo) label.
```

> To see existing requests or request support for additional architectures, use the [![](https://img.shields.io/github/labels/microsoft/go/Area-Nocgo)](https://github.com/microsoft/go/labels/Area-Nocgo) label.

> ## How the Backend is Selected

Reviewer: ```suggestion
## How the backend is selected
```

If cgo is enabled (e.g., `CGO_ENABLED=1`), the cgo-based OpenSSL backend is always used.
The cgo-less backend is only used when cgo is disabled **and** you're on a supported architecture.

> This means if you enable cgo for other reasons (e.g., linking your own C library or cross-compiling to an unsupported architecture), the crypto backend will use the cgo-based implementation.

Reviewer: > cross-compiling to an unsupported architecture

This still seems muddy to me... unsupported by whom? Why is cross-compilation significant vs. native compilation?

A related scenario that might be worth documenting is the idea that if you want to compile for an architecture not supported by the cgoless Linux backend, you can do so by enabling cgo. (Maybe it can replace this scenario.)

We could also mention that it might be enabled by default on your platform.

```suggestion
This means if cgo is enabled by default on your platform or you enable cgo for other reasons (e.g., linking your own C library), the crypto backend will use the cgo-based implementation.
You can also intentionally enable cgo to successfully build for a platform without cgo-less OpenSSL support.
```

> ## Experimental Status
> 
> While `systemcrypto` is a fully supported `GOEXPERIMENT` value (it is not "experimental"), `ms_nocgo_opensslcrypto` **is** experimental in Go 1.26 and may have limitations.

Reviewer: I think this makes more sense as a note, earlier. I would expect a section like this (if we have it) to be more detailed about why we're releasing it as an experiment first, and what our criteria will be for making it mainstream and calling it fully supported.

```suggestion
```
